proposals/0370-modifiers.rst

```diff
@@ -54,37 +54,40 @@ Proposed Change Specification
    the Haskell 2010 Report for all BNF syntax, and recall that we use braces
    to denote "zero or more")::
 
-     type     ::= btype [ {modifier} -> type ]
-     prefix%  ::= '%'    -- only in prefix position
-     modifier ::= prefix% atype
+     type          ::= btype [ {modifier} -> type ]
+     prefix%       ::= '%'    -- only in prefix position
+     modifier      ::= prefix% atype
+     modifiers     ::= {modifier}
+     declModifiers ::= [ modifier [ ';' ] [ declModifiers ] ]
 
 3. With ``-XModifiers``, introduce modifier syntax on types as follows::
 
-     btype    ::= {modifier} atype | btype atype
+     btype    ::= modifiers atype | btype atype
 
 4. With ``-XModifiers``, introduce modifier syntax on the term level as follows::
 
-     fexp     ::= {modifier} aexp | fexp aexp
+     fexp     ::= modifiers aexp | fexp aexp
 
 5. With ``-XModifiers``, introduce modifier syntax in patterns as follows::
 
-     lpat     ::= {modifier} lpat | ...
+     lpat     ::= modifiers lpat | ...
 
 6. With ``-XModifiers``, introduce modifier syntax on record field declarations as follows::
 
-     fielddecl ::= vars {modifier} '::' (type | '!' atype)
+     fielddecl ::= vars modifiers '::' (type | '!' atype)
 
 7. With ``-XModifiers``, introduce modifier syntax on top-level declarations as
    follows::
 
-     topdecl ::= {modifier} [ ';' ] 'type' simpletype '=' type
-             |   {modifier} [ ';' ] 'data' [context '=>'] simpletype ['=' constrs] [deriving]
-             |   {modifier} [ ';' ] 'newtype' [context '=>'] simpletype = newconstr [deriving]
-             |   {modifier} [ ';' ] 'class' [scontext '=>'] tycls tyvar ['where' cdecls]
-             |   {modifier} [ ';' ] 'instance' [scontext '=>'] qtycls inst ['where' idecls]
-             |   {modifier} [ ';' ] 'default' '(' type1 ',' ... ',' typen ')'
-             |   {modifier} [ ';' ] 'foreign' fdecl
-             |   {modifier} ';' decl
+     topdecl ::= declModifiers 'type' simpletype '=' type
+             |   declModifiers 'data' [context '=>'] simpletype ['=' constrs] [deriving]
+             |   declModifiers 'newtype' [context '=>'] simpletype = newconstr [deriving]
+             |   declModifiers 'type' 'data' ...
+             |   declModifiers 'class' [scontext '=>'] tycls tyvar ['where' cdecls]
+             |   declModifiers 'instance' [scontext '=>'] qtycls inst ['where' idecls]
+             |   declModifiers 'default' '(' type1 ',' ... ',' typen ')'
+             |   declModifiers 'foreign' fdecl
+             |   declModifiers ';' decl
```

Comment:
I guess this isn't technically the right syntax, because it allows two semis between the modifier and the declaration:

    %One ;; foo :: Int

I can fix this if `declModifiers` looks like it's going to be accepted.